### ISSUE TITLE
so many goroutine

### ISSUE THREAD

**Issue by lampnick** [apache/rocketmq-client-go#754]
The issue tracker is **ONLY** used for the go client (feature request of RocketMQ need to follow [RIP process](https://github.com/apache/rocketmq/wiki/RocketMQ-Improvement-Proposal)). Keep in mind, please check whether there is an existing same report before your raise a new one.

Alternately (especially if your communication is not a bug report), you can send mail to our [mailing lists](http://rocketmq.apache.org/about/contact/). We welcome any friendly suggestions, bug fixes, collaboration, and other improvements.

Please ensure that your bug report is clear and that it is complete. Otherwise, we may be unable to understand it or to reproduce it, either of which would prevent us from fixing the bug. We strongly recommend the report(bug report or feature request) could include some hints as to the following:

**BUG REPORT**  
1. Please describe the issue you observed:

    - What did you do (The steps to reproduce)?
       only subscribe consumer
    - What did you expect to see?
      few goroutine
    - What did you see instead?
      65394 goroutine
2. Please tell us about your environment:

     - What is your OS?
       alpine
     - What is your client version?
       v2.1.0
     - What is your RocketMQ version?
        aliyun 铂金版
3. Other information (e.g. detailed explanation, logs, related issues, suggestions on how to fix, etc):
```
goroutine profile: total 65394
51589 @ 0x432fc0 0x4078c8 0x40789e 0x40758b 0xc62a41 0xc62a12 0xc6bc2f 0xb4f40d 0x461511
#	0xc62a40	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*processQueue).getMessages+0x50		/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/process_queue.go:273
#	0xc62a11	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*pushConsumer).consumeMessageCurrently+0x21	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/push_consumer.go:955
#	0xc6bc2e	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*pushConsumer).pullMessage.func1+0x7e		/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/push_consumer.go:531
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c				/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

5469 @ 0x432fc0 0x42dd8a 0x42d355 0x4cfe85 0x4d0dff 0x4d0de1 0x5b55ef 0x5ca7a8 0x4ac897 0xc32a0a 0xc329c7 0xc34413 0xb4f40d 0x461511
#	0x42d354	internal/poll.runtime_pollWait+0x54														/usr/local/go/src/runtime/netpoll.go:184
#	0x4cfe84	internal/poll.(*pollDesc).wait+0x44														/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x4d0dfe	internal/poll.(*pollDesc).waitRead+0x1ce													/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x4d0de0	internal/poll.(*FD).Read+0x1b0															/usr/local/go/src/internal/poll/fd_unix.go:169
#	0x5b55ee	net.(*netFD).Read+0x4e																/usr/local/go/src/net/fd_unix.go:202
#	0x5ca7a7	net.(*conn).Read+0x67																/usr/local/go/src/net/net.go:184
#	0x4ac896	io.ReadAtLeast+0x86																/usr/local/go/src/io/io.go:310
#	0xc32a09	io.ReadFull+0x119																/usr/local/go/src/io/io.go:329
#	0xc329c6	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote.(*remotingClient).receiveResponse+0xd6	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote/remote_client.go:158
#	0xc34412	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote.(*remotingClient).connect.func1+0x32	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote/remote_client.go:133
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c				/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

3268 @ 0x432fc0 0x442bfb 0xc42b15 0xc4912a 0xb4f40d 0x461511
#	0xc42b14	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*traceDispatcher).process+0x364	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/trace.go:328
#	0xc49129	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*traceDispatcher).Start.func1+0x29	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/trace.go:292
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

1344 @ 0x432fc0 0x442bfb 0xc31bbd 0xc32116 0xc37429 0xc5f507 0xc6a9b5 0x461511
#	0xc31bbc	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote.(*ResponseFuture).waitResponse+0xbc	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote/future.go:61
#	0xc32115	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote.(*remotingClient).InvokeSync+0x275	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/remote/remote_client.go:83
#	0xc37428	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).PullMessage+0xd8			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:637
#	0xc5f506	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*pushConsumer).pullMessage+0x1276		/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/push_consumer.go:721
#	0xc6a9b4	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*pushConsumer).Start.func1.1.1+0x34		/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/push_consumer.go:156

559 @ 0x432fc0 0x442bfb 0xc457dc 0xb4f40d 0x461511
#	0xc457db	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).Start.func1.1+0x14b	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:307
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

559 @ 0x432fc0 0x442bfb 0xc45a48 0xb4f40d 0x461511
#	0xc45a47	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).Start.func1.2+0x137	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:331
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

559 @ 0x432fc0 0x442bfb 0xc46176 0xb4f40d 0x461511
#	0xc46175	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).Start.func1.4+0x165	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:387
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

559 @ 0x432fc0 0x442bfb 0xc463cb 0xb4f40d 0x461511
#	0xc463ca	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).Start.func1.5+0x11a	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:403
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

558 @ 0x432fc0 0x442bfb 0xc45d40 0xb4f40d 0x461511
#	0xc45d3f	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal.(*rmqClient).Start.func1.3+0x15f	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/internal/client.go:355
#	0xb4f40c	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive.WithRecover+0x4c			/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/primitive/base.go:96

510 @ 0x432fc0 0x442bfb 0xc6aac9 0x461511
#	0xc6aac8	xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer.(*pushConsumer).Start.func1.1+0xf8	/usr/local/go/src/xxxxxxxxxxxx/vendor/github.com/apache/rocketmq-client-go/v2/consumer/push_consumer.go:153

170 @ 0x432fc0 0x42dd8a 0x42d355 0x4cfe85 0x4d0dff 0x4d0de1 0x5b55ef 0x5ca7a8 0x6edad4 0x56af53 0x56b0bf 0x6f37c0 0x461511
#	0x42d354	internal/poll.runtime_pollWait+0x54		/usr/local/go/src/runtime/netpoll.go:184
#	0x4cfe84	internal/poll.(*pollDesc).wait+0x44		/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x4d0dfe	internal/poll.(*pollDesc).waitRead+0x1ce	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x4d0de0	internal/poll.(*FD).Read+0x1b0			/usr/local/go/src/internal/poll/fd_unix.go:169
#	0x5b55ee	net.(*netFD).Read+0x4e				/usr/local/go/src/net/fd_unix.go:202
#	0x5ca7a7	net.(*conn).Read+0x67				/usr/local/go/src/net/net.go:184
#	0x6edad3	net/http.(*connReader).Read+0xf3		/usr/local/go/src/net/http/server.go:796
#	0x56af52	bufio.(*Reader).fill+0x102			/usr/local/go/src/bufio/bufio.go:100
#	0x56b0be	bufio.(*Reader).Peek+0x4e			/usr/local/go/src/bufio/bufio.go:138
#	0x6f37bf	net/http.(*conn).serve+0x9df			/usr/local/go/src/net/http/server.go:1944

44 @ 0x432fc0 0x42dd8a 0x42d355 0x4cfe85 0x4d0dff 0x4d0de1 0x5b55ef 0x5ca7a8 0x70e415 0x56af53 0x56b0bf 0x70f0c6 0x461511
#	0x42d354	internal/poll.runtime_pollWait+0x54		/usr/local/go/src/runtime/netpoll.go:184
#	0x4cfe84	internal/poll.(*pollDesc).wait+0x44		/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x4d0dfe	internal/poll.(*pollDesc).waitRead+0x1ce	/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x4d0de0	internal/poll.(*FD).Read+0x1b0			/usr/local/go/src/internal/poll/fd_unix.go:169
#	0x5b55ee	net.(*netFD).Read+0x4e				/usr/local/go/src/net/fd_unix.go:202
#	0x5ca7a7	net.(*conn).Read+0x67				/usr/local/go/src/net/net.go:184
#	0x70e414	net/http.(*persistConn).Read+0x74		/usr/local/go/src/net/http/transport.go:1758
#	0x56af52	bufio.(*Reader).fill+0x102			/usr/local/go/src/bufio/bufio.go:100
#	0x56b0be	bufio.(*Reader).Peek+0x4e			/usr/local/go/src/bufio/bufio.go:138
#	0x70f0c5	net/http.(*persistConn).readLoop+0x1d5		/usr/local/go/src/net/http/transport.go:1911

43 @ 0x432fc0 0x442bfb 0x710b63 0x461511
#	0x710b62	net/http.(*persistConn).writeLoop+0x122	/usr/local/go/src/net/http/transport.go:2210
```
**FEATURE REQUEST**

1. Please describe the feature you are requesting.

2. Provide any additional detail on your proposed use case for this feature.

2. Indicate the importance of this issue to you (blocker, must-have, should-have, nice-to-have). Are you currently using any workarounds to address this issue?

4. If there are some sub-tasks using -[] for each subtask and create a corresponding issue to map to the sub task:

    - [sub-task1-issue-number](example_sub_issue1_link_here): sub-task1 description here, 
    - [sub-task2-issue-number](example_sub_issue2_link_here): sub-task2 description here,
    - ...

**Comment by wenfengwang** [comment]
@lampnick Thank your report, Could you tell me how to reproduce this problem?

**Comment by lampnick** [comment]
> @lampnick Thank your report, Could you tell me how to reproduce this problem?

1、RocketMQ broker topic has 32 queue.
2、Only have 4 client of one group.
3、Send 5000 messages  every hour average.
4、Maybe it's appear after two days.

**Comment by lampnick** [comment]
> > @lampnick Thank your report, Could you tell me how to reproduce this problem?
> 
> 1、RocketMQ broker topic has 32 queue. 2、Only have 4 client of one group. 3、Send 5000 messages every hour average. 4、Maybe it's appear after two days.

The problem code in rocketmq-client-go/v2/consumer/process_queue.go:292
func (pq *processQueue) getMessages() []*primitive.MessageExt {
	select {
	case <-pq.closeChan:// This line is ours simple solution.
		return nil
	case result := <-pq.msgCh://this line will block.
		return result
	}
}

**Comment by lampnick** [comment]
The better solution is to close pq.msgCh when needed.

**Comment by UnderTreeTech** [comment]
Any updates? @wenfengwang

**Comment by bilinxing** [comment]
我也发现了，groutine占用很高，这个问题什么时候能解决？

**Comment by bilinxing** [comment]
I also found that the groutine occupancy is very high, when will this problem be solved?

**Comment by ShannonDing** [comment]
@lampnick the goroutine leak seems fixed in your other PR #811 which supports pulling msg with tag , could you pls create a new PR to fix the goroutine leak separately.
